README.md
# Kafka sniffer

Initially, this tool was build for internal usage: we had a shared Kafka cluster with thousands topics and hundreds
clients (without any auth) and wanted to migrate all this stuff to the new cluster. For this reason, we needed to know
what mircoservices are still connected to the old cluster and what topics this service uses.

This sniffer can:

- detect active connections to Kafka Broker and can say who is producer and who is consumer
- detect topics to which producers trying to write / consumers trying to read
- expose IPs, request kind and topic as Prometheus metrics

Kafka protocol: https://kafka.apache.org/protocol

## changes

1. 2022-02-24 add `/client` API to get the result of client statistics, See the [demo](#demo1client).
2. 2022-02-24 Print clients of `kafka.ProduceRequest` and `kafka.FetchRequest`, see the [demo](#demo1).

## example

Example:

```
// Run simple producer who writes to topic "mytopic"
go run cmd/producer/main.go -brokers 127.0.0.1:9092

// Run sniffer on net iface (loopback or usually, eth0)
go run cmd/sniffer/main.go -i=lo0

// OR with debug info:
go run cmd/sniffer/main.go -i=lo0 -assembly_debug_log=false
```

Example output:

```
2020/05/16 16:25:49 got request, key: 0, version: 0, correlationID: 132, clientID: sarama
2020/05/16 16:25:49 client 127.0.0.1:60423 wrote to topic mytopic
2020/05/16 16:25:54 got request, key: 0, version: 0, correlationID: 133, clientID: sarama
2020/05/16 16:25:54 client 127.0.0.1:60423 wrote to topic mytopic
2020/05/16 16:25:59 got request, key: 0, version: 0, correlationID: 134, clientID: sarama
2020/05/16 16:25:59 client 127.0.0.1:60423 wrote to topic mytopic
2020/05/16 16:26:04 got request, key: 0, version: 0, correlationID: 135, clientID: sarama
2020/05/16 16:26:04 client 127.0.0.1:60423 wrote to topic mytopic
2020/05/16 16:26:05 got EOF - stop reading from stream
```

## Run as a Docker container

```
docker build . -t kafka-sniffer
docker run --rm --network host kafka-sniffer:latest -i lo0
```

## Run Kafka in docker (bitnami Kafka + Zookeeper)

```
docker-compose ./etc/docker-compose.yml up
```

## Run Kafka in minicube (Strimzi Kafka Operator)

```yaml
kubectl create namespace kafka
kubectl apply -f 'https://strimzi.io/install/latest?namespace=kafka' -n kafka
kubectl apply -f https://strimzi.io/examples/latest/kafka/kafka-persistent-single.yaml -n kafka
kubectl wait kafka/my-cluster --for=condition=Ready --timeout=300s -n kafka
```

Send & receive messages

```yaml
kubectl -n kafka run kafka-producer -ti --image=strimzi/kafka:0.17.0-kafka-2.4.0 --rm=true --restart=Never -- bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9092 --topic my-topic
kubectl -n kafka run kafka-consumer -ti --image=strimzi/kafka:0.17.0-kafka-2.4.0 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --topic my-topic --from-beginning
```

Port-forwarding for local development:

```yaml
kubectl port-forward service/my-cluster-kafka-brokers 9092
```

And probably you'll need to add this row to `/etc/hosts`

```yaml
127.0.0.1   my-cluster-kafka-0.my-cluster-kafka-brokers.kafka.svc
```

## demo1

```sh
# kafka-sniffer
2022/02/24 13:34:42 starting capture on interface "eth0"
2022/02/24 13:34:42 conn: 192.1.1.15:61285 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [dev-logcenter], correlationID: 117377425, clientID: sarama
2022/02/24 13:34:42 conn: 192.1.1.15:37953 -> 192.1.1.14:9092 type: *kafka.ProduceRequest topics: [dev-metrics], correlationID: 6003063, clientID: sarama
2022/02/24 13:34:42 conn: 192.1.1.11:24717 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [dev-metrics], correlationID: 196489671, clientID: sarama
2022/02/24 13:34:42 conn: 192.1.1.7:37233 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [__consumer_offsets], correlationID: 247189, clientID: consumer-KMOffsetCache-cmak-548974c6c4-sxvgt-1723
2022/02/24 13:34:42 conn: 192.1.6.17:51404 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [dev-ids], correlationID: 6716609, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.9.23:36866 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [bq_disaster_recovery], correlationID: 623626, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.7:34038 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [agent_transaction], correlationID: 12480162, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.4:55214 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [dev-cloudSignLogServer], correlationID: 3341672, clientID: 2428545257036493
2022/02/24 13:34:42 conn: 192.1.1.12:6267 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [judicial_disaster], correlationID: 9009620, clientID: consumer-2
2022/02/24 13:34:42 conn: 192.1.1.11:33378 -> 192.1.1.14:9092 type: *kafka.ProduceRequest topics: [dev-gateway], correlationID: 10948681, clientID: producer-1
2022/02/24 13:34:42 conn: 192.1.1.12:9195 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [judicial-2tripartite], correlationID: 9011202, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.12:41426 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [agent_count_transaction], correlationID: 194647, clientID: consumer-11
2022/02/24 13:34:42 conn: 192.1.1.11:22615 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [ids-message-record-1], correlationID: 8999184, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.12:20394 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [count_transaction_pro], correlationID: 3240311, clientID: consumer-11
2022/02/24 13:34:42 conn: 192.1.1.12:7273 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [transaction_pro], correlationID: 3240395, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.4:6654 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [count_transaction], correlationID: 572423, clientID: consumer-1
2022/02/24 13:34:42 conn: 192.1.1.4:48249 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [transaction], correlationID: 8692411, clientID: consumer-11
2022/02/24 13:34:42 conn: 192.1.9.23:33500 -> 192.1.1.14:9092 type: *kafka.FetchRequest topics: [verif_supplement_file_v1], correlationID: 117992, clientID: consumer-2
```

### demo1client

![image](https://user-images.githubusercontent.com/1940588/155554492-f319b5bb-b50c-47e5-b19d-5e1a2d23e119.png)

```sh
# gurl :9870/client
Conn-Session: 127.0.0.1:44828->127.0.0.1:9870 (reused: false, wasIdle: false, idle: 0s)
GET /client? HTTP/1.1
Host: 127.0.0.1:9870
Accept: application/json
Accept-Encoding: gzip, deflate
Content-Type: application/json
Gurl-Date: Thu, 24 Feb 2022 06:30:03 GMT
User-Agent: gurl/1.0.0


HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Thu, 24 Feb 2022 06:30:03 GMT

[
  {
    "Start": "2022-02-24T14:29:19.049701376+08:00",
    "Client": "192.1.6.17:51404",
    "ReqType": "*kafka.FetchRequest",
    "ClientID": "consumer-1",
    "Requests": 89,
    "BytesRead": 7387,
    "Topics": [
      "dev-ids"
    ]
  },
  {
    "Start": "2022-02-24T14:29:19.025437041+08:00",
    "Client": "192.1.8.12:6267",
    "ReqType": "*kafka.FetchRequest",
    "ClientID": "consumer-2",
    "Requests": 89,
    "BytesRead": 7031,
    "Topics": [
      "judicial_disaster"
    ]
  },
  {
    "Start": "2022-02-24T14:29:20.301435997+08:00",
    "Client": "192.1.6.15:56324",
    "ReqType": "*kafka.ProduceRequest",
    "ClientID": "sarama",
    "Requests": 309,
    "BytesRead": 123625,
    "Topics": [
      "dev-metrics"
    ]
  },
  {
    "Start": "2022-02-24T14:29:20.84427283+08:00",
    "Client": "192.1.6.4:54598",
    "ReqType": "*kafka.ProduceRequest",
    "ClientID": "sarama",
    "Requests": 283,
    "BytesRead": 113472,
    "Topics": [
      "dev-metrics"
    ]
  }
]
```

## captures TCP/IP traffic and keeps conns track

```sh
# kafka-sniffer -flow
2022/02/28 16:37:42 starting capture on interface "eth0"
2022/02/28 16:37:42 start to captures TCP/IP traffic and keeps conns track, using bpf "tcp and dst port 9092" on device "eth0"
2022/02/28 16:37:42 serving metrics and api on :9870
```

```sh
# gurl :9870/flow -r
Conn-Session: 127.0.0.1:36876->127.0.0.1:9870 (reused: false, wasIdle: false, idle: 0s)
GET /flow? HTTP/1.1
Host: 127.0.0.1:9870
Accept: application/json
Accept-Encoding: gzip, deflate
Content-Type: application/json
Gurl-Date: Mon, 28 Feb 2022 08:39:05 GMT
User-Agent: gurl/1.0.0


HTTP/1.1 200 OK
Content-Type: application/json
Date: Mon, 28 Feb 2022 08:39:05 GMT

{"Total":254,"Flows":[{"Src":"192.1.8.4:1592","Dst":"192.1.8.14:9092","PayloadSum":52,"Closed":false,"Update":"2022-02-28T16:38:44.274237328+08:00"},{"Src":"192.1.8.4:37234","Dst":"192.1.8.14:9092","PayloadSum":52,"Closed":false,"Update":"2022-02-28T16:38:44.275974959+08:00"},{"Src":"192.1.8.4:55852","Dst":"192.1.8.14:9092","PayloadSum":188647,"Closed":false,"Update":"2022-02-28T16:39:02.667679646+08:00"},{"Src":"192.1.8.12:43326","Dst":"192.1.8.14:9092","PayloadSum":225143,"Closed":false,"Update":"2022-02-28T16:39:03.678359898+08:00"},{"Src":"192.1.8.4:46099","Dst":"192.1.8.14:9092","PayloadSum":57,"Closed":false,"Update":"2022-02-28T16:38:44.362084935+08:00"},{"Src":"192.1.8.4:58497","Dst":"192.1.8.14:9092","PayloadSum":2430,"Closed":false,"Update":"2022-02-28T16:39:03.503698159+08:00"}]}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v0.0.1] - 2020-05-25

### Added

- Base sniffer functionality of capturing kafka producer/consumer and topics relations.


cmd/kafka-sniffer/main.go
package main

import (
	"flag"
	"github.com/bingoohuang/kafka-sniffer/stream/flowd"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/bingoohuang/kafka-sniffer/metrics"
	"github.com/bingoohuang/kafka-sniffer/stream"

	"github.com/google/gopacket"
	"github.com/google/gopacket/examples/util"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
	"github.com/google/gopacket/tcpassembly"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	pType      = flag.String("t", "", "req types filter, e.g. FetchRequest, ProduceRequest")
	iface      = flag.String("i", "eth0", "Interface to get packets from")
	bpf        = flag.String("bpf", "tcp and dst port 9092", "BPF expr")
	snaplen    = flag.Int("snap", 16<<10, "SnapLen for pcap packet capture")
	verbose    = flag.Bool("v", false, "Logs every packet in great detail")
	rwPrint    = flag.Bool("s", true, "Print the read and write clients")
	connTrack  = flag.Bool("flow", false, "Captures TCP/IP traffic and keeps conns track")
	listenAddr = flag.String("addr", ":9870", "Address on which sniffer listen the requests, e.g. :9870")
	expireTime = flag.Duration("metrics.expire-time", 5*time.Minute, "Expiration time of metric.")

	printJsonDuration = flag.Duration("p", 0, "Print the request json")

	clientStat *stream.ClientStat
)

func init() {
	flag.Parse()
	log.SetOutput(os.Stdout)
}

func main() {
	defer util.Run()()

	log.Printf("starting capture on interface %q", *iface)

	if *connTrack {
		handler, err := flowd.RunNetworkAnalyzer(*iface, *bpf, int32(*snaplen))
		if err != nil {
			log.Fatalf("failed to create network analyzer, err: %v", err)
		}
		log.Printf("start to captures TCP/IP traffic and keeps conns track, using bpf %q on device %q", *bpf, *iface)
		http.HandleFunc("/flow", handler)
	} else {
		go handlerKafka()
	}

	runTelemetry()
}

func handlerKafka() {
	// Set up pcap packet capture
	handle, err := pcap.OpenLive(*iface, int32(*snaplen), true, pcap.BlockForever)
	if err != nil {
		panic(err)
	}

	if err := handle.SetBPFFilter(*bpf); err != nil {
		panic(err)
	}

	// init metrics storage
	metricsStorage := metrics.NewStorage(prometheus.DefaultRegisterer, *expireTime)

	// Set up assembly
	var f tcpassembly.StreamFactory
	if *rwPrint {
		xf := stream.NewKafkaClientPrintStreamFactory(*printJsonDuration, *pType)
		clientStat = xf.ClientStat
		f = xf
	} else {
		f = stream.NewKafkaStreamFactory(metricsStorage, *verbose)
	}

	streamPool := tcpassembly.NewStreamPool(f)
	assembler := tcpassembly.NewAssembler(streamPool)

	// Auto-flushing connection state to get packets
	// without waiting SYN
	assembler.MaxBufferedPagesTotal = 1000
	assembler.MaxBufferedPagesPerConnection = 1

	if *verbose {
		log.Println("reading in packets")
	}

	// Read in packets, pass to assembler.
	packets := gopacket.NewPacketSource(handle, handle.LinkType()).Packets()
	ticker := time.Tick(time.Minute)

	for {
		select {
		case p := <-packets:
			if *verbose {
				log.Println(p)
			}

			n := p.NetworkLayer()
			t := p.TransportLayer()
			if n == nil || t == nil || t.LayerType() != layers.LayerTypeTCP {
				if *verbose {
					log.Println("Unusable packet")
				}
				continue
			}

			tcp := t.(*layers.TCP)
			assembler.AssembleWithTimestamp(n.NetworkFlow(), tcp, p.Metadata().Timestamp)

		case <-ticker:
			// Every minute, flush connections that haven't seen activity in the past 2 minutes.
			assembler.FlushOlderThan(time.Now().Add(time.Minute * -2))
			if *verbose {
				log.Println("---- FLUSHING ----")
			}
		}
	}
}

func runTelemetry() {
	log.Printf("serving metrics and api on %s\n", *listenAddr)

	http.Handle("/metrics", promhttp.Handler())
	if clientStat != nil {
		http.Handle("/client", stream.ServeClientStatHandler(clientStat))
	}

	if err := http.ListenAndServe(*listenAddr, nil); err != nil {
		panic(err)
	}
}
